service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/CbsReceiverLinkHandler.java

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.azure.sdk.iot.service.transport.amqps;
+
+import com.microsoft.azure.sdk.iot.deps.transport.amqp.AmqpsMessage;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.qpid.proton.amqp.messaging.Properties;
+import org.apache.qpid.proton.amqp.messaging.Released;
+import org.apache.qpid.proton.amqp.transport.DeliveryState;
+import org.apache.qpid.proton.engine.Delivery;
+import org.apache.qpid.proton.engine.Event;
+import org.apache.qpid.proton.engine.Receiver;
+
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Every SAS token based authentication over AMQP requires a CBS session with a sender and receiver link. This
+ * class defines the receiver link which receives authentication status codes corresponding to each authentication attempt.
+ */
+@Slf4j
+public final class CbsReceiverLinkHandler extends ReceiverLinkHandler
+{
+    private final static String APPLICATION_PROPERTY_STATUS_CODE = "status-code";
+    private final static String APPLICATION_PROPERTY_STATUS_DESCRIPTION = "status-description";
+
+    private static final String RECEIVER_LINK_ENDPOINT_PATH = "$cbs";
+    private static final String RECEIVER_LINK_TAG_PREFIX = "cbs-receiver";
+    private static final String LINK_TYPE = "cbs";
+
+    private UUID authenticationMessageCorrelationId;
+    private final AuthenticationMessageCallback authenticationMessageCallback;
+
+    CbsReceiverLinkHandler(Receiver receiver, AuthenticationMessageCallback authenticationMessageCallback, LinkStateCallback linkStateCallback)
+    {
+        super(receiver, UUID.randomUUID().toString(), linkStateCallback);
+        this.receiverLinkTag = RECEIVER_LINK_TAG_PREFIX;
+        this.receiverLinkAddress = RECEIVER_LINK_ENDPOINT_PATH;
+        this.authenticationMessageCallback = authenticationMessageCallback;
+    }
+
+    protected void setAuthenticationMessageCorrelationId(UUID authenticationMessageCorrelationId)
+    {
+        this.authenticationMessageCorrelationId = authenticationMessageCorrelationId;
+    }
+
+    static String getCbsTag()
+    {
+        return RECEIVER_LINK_TAG_PREFIX;
+    }
+
+    @Override
+    public String getLinkInstanceType()
+    {
+        return LINK_TYPE;
+    }
+
+    @Override
+    public void onDelivery(Event event)
+    {
+        Delivery delivery = event.getDelivery();
+        log.trace("Received a message on the CBS receiver link");
+        handleCBSResponseMessage();
+        delivery.free();
```

Comment:
Should delivery.free() be in a try finally block? What does free do?

Reply:
In practice, it calls delivery.settle();

From the proton-j docs:

```java
    /**
     * Settles this delivery.
     *
     * **Causes the delivery to be removed from the connection's work list** (see {@link Connection#getWorkHead()}).
     * If this delivery is its link's current delivery, the link's current delivery pointer is advanced.
     */
```

Reply:
In this case, the free() call will always happen because none of the lines of code before it throw any exceptions, so it isn't necessary to put it in a finally block

Reply:
Ok. The term seemed like free some memory like in Dispose :P

Reply:
Kind of, yeah